Implemented tests for ConfigDispatcher

petarnv opened the pull request:
There are 24 tests, 5 of which are ignored, because I think the behavior of the ConfigDispatcher in them is unexpected:
    1) **Incorrect Overriding** - I think that the configuration should be updated with the last modified value for a specific property. In ConfigDispatcher files are processed as listFiles() method ordered them and this order is never guaranteed(https://docs.oracle.com/javase/7/docs/api/java/io/File.html#listFiles()), i.e. the order can be alphabetical or some other order, which may cause the latest value for a property to be overridden by another value of the same property.
    2) **No Update On Delete** - On delete the file should be processed and the configurations should be updated, but in ConfigDispatcher, when file is deleted, it is only processed, but not updated. An update happens only when new content is added, but not when it is deleted.

As we have seen from https://github.com/eclipse/smarthome/pull/1597 and https://github.com/eclipse/smarthome/pull/2074, the API for files watching is platform dependent. These proposed tests are behaving exactly the same way on Windows and MAC.

petarnv commented:
I saw that the ip-validation failed. I will fix my CLA.

maggu2810 commented:
> 1) Incorrect Overriding - I think that the configuration should be updated with the last modified value for a specific property. In ConfigDispatcher files are processed as listFiles() method ordered them and this order is never guaranteed(https://docs.oracle.com/javase/7/docs/api/java/io/File.html#listFiles()), i.e. the order can be alphabetical or some other order, which may cause the latest value for a property to be overridden by another value of the same property.

I assume we are able to sort the list returned by listFiles() to a specific order.
We can order it by the file name, the last modification time, etc.
It need at least be defined.
The config dispatcher extends the AbstractWatchService so at runtime a modification will result on a reload of the last modified file.
Doesn't make a sort by modification time sense here?

> 2) No Update On Delete - On delete the file should be processed and the configurations should be updated, but in ConfigDispatcher, when file is deleted, it is only processed, but not updated. An update happens only when new content is added, but not when it is deleted.

Looking at the sources it seems that only ENTRY_CREATE and ENTRY_MODIFY are handled and delete is ignored at all.

petarnv commented:
> I assume we are able to sort the list returned by listFiles() to a specific order.
We can order it by the file name, the last modification time, etc.
It need at least be defined.
The config dispatcher extends the AbstractWatchService so at runtime a modification will result on a reload of the last modified file.
Doesn't make a sort by modification time sense here?

Yes - a modification will result in a reload of the last modified file and in this case the configuration will be updated properly and no sort is even needed. But that is not the the thing I test in the test methods from https://github.com/eclipse/smarthome/issues/2109. After activated, the ConfigDispatcher reads all the configuration files and processes them. And in the ignored tests I want to test precisely the processing after activation of the ConfigDispatcher, not the processing on modification of a file (I have another test that tests the processing on modification of a file). And the problem is that when the ConfigDispatcher is started, it processes files as listFiles() ordered them (not guaranteed order). So, if any change happened in the content of these files before the activation of the ConfigDispatcher, а modification event will not be handled and all the files (modified or not) will be processed as listFiles() ordered them and the configuration might not be updated properly.

> Looking at the sources it seems that only ENTRY_CREATE and ENTRY_MODIFY are handled and delete is ignored at all.

Now I see that my explanation can be misleading. What I meant was that in the  ConfigDispatcher, when the content of a file(not the file itself) is deleted, it is only processed, but not updated. It is processed because when we delete the content of the file this is considered ENTRY_MODIFY, not ENTRY_DELETE. And as I have said - an update happens only when new content is added, but not when it is deleted.

maggu2810 commented:
>> I assume we are able to sort the list returned by listFiles() to a specific order.
We can order it by the file name, the last modification time, etc.

> And the problem is that when the ConfigDispatcher is started, it processes files as listFiles() ordered them (not guaranteed order). 

As written above, we can sort and process the files that are returned by listFiles() to grant that also in this case the one with the last modification timestamp wins. Doesn't it solve your non deterministic behaviour if we state "also on startup the files are read in order of the modification timestamp"?

petarnv commented:
I have added a sort by modification time of the files that are read at the activation of the ConfigDispatcher, so that the last modified file is processed last. I also removed the `@Ignore` tag in the tests, related to the incorrect overriding, because they are not failing after the change.

maggu2810 commented:
@petarnv Sorry about the delay. Can you give me a short summary what is missing? Are there still ignored tests?

petarnv commented:
@maggu2810, there is still one ignored test left, because when a content of a file is deleted, the configuration is not updated properly (it should be updated without the deleted from the file values), but I am working on fixing this problem.

maggu2810 commented:
Thanks. If there is no urgent reason to merge this PR (with ignoring one test) now, I will wait for the fix, so we do not need to ignore a test. Okay?

petarnv commented:
Okay

petarnv commented:
I have updated the PR. I have fixed the problem with the configuration not being updated, when content of a file is deleted and I have added a couple more tests, that show the fix is really working. The new tests, as well as the old ones, are behaving exactly the same way on Windows, Linux and MAC. There are no more ignored tests.

kaikreuzer commented:
Sorry to come late to the party.

> Looking at the sources it seems that only ENTRY_CREATE and ENTRY_MODIFY are handled and delete is ignored at all.

This is actually done this way on purpose. With the current implementation, the behavior is identical whether your runtime is up or down at the moment when you remove a file.
With your implementation, the parameters are removed from ConfigAdmin if you delete the file while the runtime is up. But if you delete it while the runtime is down and you start it up, the settings in ConfigAdmin are kept (because you have no clue that this file ever existed).

So while it is not ideal, it at least is consistent. The only way to get rid off settings is to set parameters to null (empty string?) - but configurations themselves can only be deleted through other mechanisms (config admin through console, starting the OGSi framework with clean option, etc.).

petarnv commented on bundles/config/org.eclipse.smarthome.config.dispatch/src/main/java/org/eclipse/smarthome/config/dispatch/internal/ConfigDispatcher.java:
```diff
@@ -38,39 +42,47 @@
 import org.slf4j.LoggerFactory;
 
 /**
- * This class provides a mean to read any kind of configuration data from
```
Sorry about this. I will revert to the original formatting, when I update the PR.

petarnv commented:
@kaikreuzer, I agree with you, that the change, related to the configuration delete would not behave the same way, if the runtime is down. I can remove it, when I update the PR. But what should I do with the tests, that would fail, if that change is removed - should I ignore them, or should I remove them.

And also, there is a second change in the ConfigDispatcher, that I have made - the sort of the files by modification time, when they are read. Do you approve this change?

kaikreuzer commented:
> But what should I do with the tests, that would fail, if that change is removed - should I ignore them, or should I remove them.

Probably ok to remove them, if they were meant for a functionality that actually isn't there.

> And also, there is a second change in the ConfigDispatcher, that I have made - the sort of the files by modification time, when they are read. Do you approve this change?

Yes, I saw this. I would consider it an edge case, because in theory, no two files should ever deal with the same configuration parameters - but IF this situation occurs, your fix indeed makes sense.

petarnv commented:
@kaikreuzer, I have updated my PR with the changes, that we discussed.

kaikreuzer commented:
Great, many thanks!
